File: lab3/src/main/java/lab3/interaction/SimpleDesigner.java
package lab3.interaction;

import lab3.functionality.*;

import java.util.ArrayList;

public class SimpleDesigner implements Designer {
    private final Reader reader;
    private final Writer writer;
    private final Player player;

    public SimpleDesigner(Reader reader, Writer writer, Player player) {
        this.reader = reader;
        this.writer = writer;
        this.player = player;
    }

    @Override
    public void interactionSession() {
        showVariants();
    }

    @Override
    public void endMessage(String reason) {
        writer.writeln(reason);
    }

    private void showVariants() {
        writer.writeln(String.format("Выберите, с чем взаимодействовать:\n1)Посмотреть состояние %s\n2)Локации\n3)Предметы", player.getName()));
        var num = reader.readInt();
        switch (num) {
            case 1 -> chooseAction(player);
            case 2 -> chooseLocation();
            case 3 -> chooseItem();
            default -> writer.writeln("Неверное число");
        }
    }

    private void chooseLocation() {
        writer.writeln(String.format("Выберите локацию:\n0)Вернуться\n1)Осмотреть %s (текущая)", player.getLocation().getName()));
        var locations = player.getLocation().getLocations();
        showNames(locations, 2);
        var ind = reader.readInt();
        switch (ind) {
            case 0:
                break;
            case 1:
                chooseAction(player.getLocation());
                break;
            default:
                try {
                    writer.writeln(player.setLocation(locations.get(ind - 2)));
                } catch (IndexOutOfBoundsException e) {
                    writer.writeln("Неправильное число");
                }
        }
    }

    private void chooseItem() {
        writer.writeln("Выберите предмет:\n0)Вернуться");
        var items = player.getLocation().getItems();
        showNames(items, 1);
        var ind = reader.readInt();
        if (ind != 0) {
            try {
                var item = items.get(ind - 1);
                chooseAction(item);
                if (item.isConsumable() && item.getActions().size() == 0) {
                    player.getLocation().removeItem(ind - 1);
                }
            } catch (IndexOutOfBoundsException e) {
                writer.writeln("Неверное число!");
            }
        }
    }

    private void chooseAction(Actionable act) {
        showDescription(act);
        writer.writeln("Выберите действие:\n0)Вернуться");
        var actions = act.getActions();
        showNames(actions, 1);
        var ind = reader.readInt();
        if (ind != 0) {
            try {
                var action = actions.get(ind - 1);
                showDescription(action);
                actions.get(ind - 1).use(player);
                if (actions.get(ind - 1).isDisposable()) {
                    try {
                        act.removeAction(ind - 1);
                    } catch (IndexOutOfBoundsException e) {
                    }
                }
            } catch (IndexOutOfBoundsException e) {
                writer.writeln("Неверное число!");
            }
        }
    }

    private void showDescription(Nameable describable) {
        var description = describable.getDescription(player);
        if (description.length() > 0) {
            writer.writeln(description);
        }
        writer.writeln("");
    }

    private void showNames(ArrayList<? extends Nameable> array, int shift) {
        for (var i = 0; i < array.size(); i++) {
            writer.writeln(String.format("%d)%s", i + shift, array.get(i).getName()));
        }
    }
}

File: lab4/src/main/kotlin/lab4/interation/SimpleDesigner.kt
package lab4.interation

import lab4.functionality.*

class SimpleDesigner(private val reader: Reader, private val writer: Writer, private val player: Player) : Designer {

    override fun interactionSession() {
        showVariants()
    }

    override fun endMessage(reason: String) {
        writer.writeln(reason)
    }

    private fun showVariants() {
        writer.writeln("Выберите, с чем взаимодействовать:\n1)Посмотреть состояние ${player.name}\n2)Локации\n3)Предметы")
        try {
            when (reader.readInt()) {
                1 -> chooseAction(player)
                2 -> chooseLocation()
                3 -> chooseItem()
                666 -> throw DemonInfiltratingTheSystem()
                else -> writer.writeln("Неверное число")
            }
        } catch (e: IndexOutOfBoundsException) {
            writer.writeln("Неверное число!")
        }
    }

    private fun chooseLocation() {
        writer.writeln("Выберите локацию:\n0)Вернуться\n1)Осмотреть ${player.location.name} (текущая)")
        val locations = player.location.locations
        showNames(locations, 2)
        when (val ind = reader.readInt()) {
            0 -> {}
            1 -> chooseAction(player.location)
            else -> {
                val loc = locations[ind - 2]
                player.location = loc
                writer.writeln(loc.enter(player))
            }
        }
    }

    private fun chooseItem() {
        writer.writeln("Выберите предмет:\n0)Вернуться")
        val items = player.location.items
        showNames(items, 1)
        val ind = reader.readInt()
        if (ind != 0) {
            val item = items[ind - 1]
            chooseAction(item)
            if (item.isConsumable && item.actions.size == 0) {
                items.removeAt(ind - 1)
            }
        }
    }

    private fun chooseAction(obj: Actionable) {
        showDescription(obj)
        writer.writeln("Выберите действие:\n0)Вернуться")
        val actions = obj.actions
        showNames(actions, 1)
        val ind = reader.readInt()
        if (ind != 0) {
            val action = actions[ind - 1]
            showDescription(action)
            actions[ind - 1].use(player, obj)
            if (actions[ind - 1].isDisposable) {
                actions.removeAt(ind - 1)
            }
        }
    }

    private fun showDescription(describable: Nameable) {
        val description = describable.description.getDescription(player)
        if (description.isNotEmpty()) {
            writer.writeln(description)
        }
        writer.writeln("")
    }

    private fun showNames(array: ArrayList<out Nameable>, shift: Int) {
        for (i in array.indices) {
            writer.writeln("${i + shift})${array[i].name}")
        }
    }
}